third_party/proton/csrc/lib/Profiler/RoctracerProfiler.cpp

```diff
@@ -0,0 +1,327 @@
+    } break;
+    case HIP_API_ID_hipExtLaunchMultiKernelMultiDevice: {
+      auto &params =
+          data->args.hipExtLaunchMultiKernelMultiDevice.launchParamsList__val;
+      name =
+          cxx_demangle(hip::getKernelNameRefByPtr(params.func, params.stream));
+    } break;
+    case HIP_API_ID_hipExtModuleLaunchKernel: {
+      auto &params = data->args.hipExtModuleLaunchKernel;
+      name = cxx_demangle(hip::getKernelNameRef(params.f));
+    } break;
+    case HIP_API_ID_hipHccModuleLaunchKernel: {
+      auto &params = data->args.hipHccModuleLaunchKernel;
+      name = cxx_demangle(hip::getKernelNameRef(params.f));
+    } break;
+    case HIP_API_ID_hipLaunchCooperativeKernel: {
+      auto &params = data->args.hipLaunchCooperativeKernel;
+      name = cxx_demangle(hip::getKernelNameRefByPtr(params.f, params.stream));
+    } break;
+    case HIP_API_ID_hipLaunchCooperativeKernelMultiDevice: {
+      auto &params = data->args.hipLaunchCooperativeKernelMultiDevice
+                         .launchParamsList__val;
+      name =
+          cxx_demangle(hip::getKernelNameRefByPtr(params.func, params.stream));
+    } break;
+    case HIP_API_ID_hipLaunchKernel: {
+      auto &params = data->args.hipLaunchKernel;
+      name = cxx_demangle(
+          hip::getKernelNameRefByPtr(params.function_address, params.stream));
+    } break;
+    case HIP_API_ID_hipModuleLaunchKernel: {
+      auto &params = data->args.hipModuleLaunchKernel;
+      name = cxx_demangle(hip::getKernelNameRef(params.f));
+    } break;
+    case HIP_API_ID_hipGraphLaunch: {
+      name = "graphLaunch";
+    } break;
+    default:;
+    }
+  }
+  return name;
+}
+
+} // namespace
+
+void RoctracerProfiler::apiCallback(uint32_t domain, uint32_t cid,
+                                    const void *callback_data, void *arg) {
+  auto [isRuntimeAPI, isDriverAPI] = matchKernelCbId(cid);
+  if (!(isRuntimeAPI || isDriverAPI)) {
+    return;
+  }
+  RoctracerProfiler &profiler =
+      dynamic_cast<RoctracerProfiler &>(RoctracerProfiler::instance());
+  if (domain == ACTIVITY_DOMAIN_HIP_API) {
+    const hip_api_data_t *data = (const hip_api_data_t *)(callback_data);
+    if (data->phase == ACTIVITY_API_PHASE_ENTER) {
+      // if (callbackData->context && roctracerState.level == 0) {
+      {
+        // Valid context and outermost level of the kernel launch
+        const char *name = kernelName(domain, cid, callback_data);
+        // roctracer::getOpString(ACTIVITY_DOMAIN_HIP_API, cid, 0);	//
+        // proper api name
+        auto scopeId = Scope::getNewScopeId();
+        auto scope = Scope(scopeId, name);
+        roctracerState.record(scope, profiler.getDataSet());
+        roctracerState.enterOp();
+
+        // Generate and Report external correlation
+        for (int it = CorrelationDomain::begin; it < CorrelationDomain::end;
+             ++it) {
+          if (externalIdMap[it].size() > 0) {
+            profiler.correlation[data->correlation_id] =
```

Comment:
This is not thread safe. `correlation` is shared in both the user thread and the activity processing thread.